sdk/core/System.BinaryData/src/BinaryData.cs

```diff
@@ -0,0 +1,286 @@
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from a string by converting
+        /// the string to bytes using the UTF-8 encoding.
+        /// </summary>
+        /// <param name="data">The string data.</param>
+        /// <returns>A value representing the UTF-8 encoding of <paramref name="data"/>.</returns>
+        public static BinaryData FromString(string data) =>
+            new BinaryData(data);
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static BinaryData FromStream(Stream stream)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+#pragma warning disable AZC0102 // Do not use GetAwaiter().GetResult().
+            return FromStreamAsync(stream, false).GetAwaiter().GetResult();
+#pragma warning restore AZC0102 // Do not use GetAwaiter().GetResult().
+        }
+
+        /// <summary>
+        /// Creates a <see cref="BinaryData"/> instance from the specified stream.
+        /// The stream is not disposed by this method.
+        /// </summary>
+        /// <param name="stream">Stream containing the data.</param>
+        /// <param name="cancellationToken">A token that may be used to cancel the operation.</param>
+        /// <returns>A value representing all of the data remaining in <paramref name="stream"/>.</returns>
+        public static Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            CancellationToken cancellationToken = default)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            return FromStreamAsync(stream, true, cancellationToken);
+        }
+
+        private static async Task<BinaryData> FromStreamAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken = default)
+        {
+            int streamLength = 0;
+            if (stream.CanSeek)
+            {
+                long longLength = stream.Length - stream.Position;
+                if (longLength > int.MaxValue)
+                {
+                    throw new ArgumentOutOfRangeException(
+                        nameof(stream),
+                        "Stream length must be less than Int32.MaxValue");
+                }
+                streamLength = (int)longLength;
+            }
+
+            ArraySegment<byte> buffer = default;
+
+            if (stream.GetType() == typeof(MemoryStream) && ((stream as MemoryStream)?.TryGetBuffer(out buffer) ?? false))
```

Comment:
Not a fan of this special-cased logic. This means that the caller can no longer reason about the ownership of data provided via `BinaryData.FromStream(Stream)`. This method should _always_ copy, or it should _never_ copy. It shouldn't exist in a quantum state.

Reply:
This was based on a suggestion from @bartonjs that we consider special-casing memoryStream for perf. I am open to implementing whatever is agreed upon.

Reply:
Talked offline - will revert this change and just always copy.

Reply:
FWIW, it's ok to special-case `MemoryStream` for perf, but this shouldn't result in a caller-visible behavioral change. An optimization which affects ownership is caller-visible and could surprise people.

An optimization which would _not_ be caller-visible would be to call `MemoryStream.ToArray`. It's more efficient than a manual copy routine, _and_ it preserves the behavior that using the `Stream`-based factory always results in an independent copy of the data.